returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,202 @@
+            # only required when data shapes change wrt. the wrapped dataset:
+            "map_outputs": {
+                "data": {"dims": [time_dim, new_data_dim]},
+            },
+        }
+    """
+
+    def __init__(
+        self,
+        dataset: Dict[str, Any],
+        map_seq: Optional[Union[Callable[[TensorDict], TensorDict]]] = None,
+        map_seq_stream: Optional[Callable[[Iterator[TensorDict]], Iterator[TensorDict]]] = None,
+        map_outputs: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ):
+        """
+        :param dataset: inner dataset to be post-processed
+        :param map_seq: post processor function operating on the single-segment level.
+            To avoid confusion on the order of how the processing functions are applied to the data, only one of
+            `map_seq` and `map_seq_stream` can be specified at a time.
+        :param map_seq_stream: post processor function operating on the multiple segment level via an iterator.
+            Allows merging multiple segments into one, or generating multiple output segments from one input segment.
+            To avoid confusion on the order of how the processing functions are applied to the data, only one of
+            `map_seq` and `map_seq_stream` can be specified at a time.
+        :param map_outputs: Type and axis specification of the outputs of the mapping functions,
+            like extern_data and model_outputs.
+            To simplify the common case when no shapes change, this value can be left unspecified. The dataset then
+            assumes the same data layout as returned by the wrapped dataset.
+            Example: `map_outputs={"data": {"dim": 42}}`
+        :param kwargs: see :class:`CachedDataset2`, :class:`Dataset`
+        """
+        super().__init__(**kwargs)
+
+        if self.seq_ordering != "default":
+            raise ValueError(f"{self}: specify seq_ordering in wrapped dataset, not in {self.__class__.__name__}")
+        if map_seq is None and map_seq_stream is None:
+            raise ValueError(f"{self}: need to either set map_seq or map_seq_stream")
+        if map_seq and map_seq_stream:
+            raise ValueError(f"{self}: cannot set both map_seq and map_seq_stream")
+
+        self._dataset_def = dataset
+        self._map_seq = map_seq
+        self._map_seq_stream = map_seq_stream
+        self._map_outputs = map_outputs
+
+        self._dataset = init_dataset(self._dataset_def, parent_dataset=self)
+        self._estimated_num_seqs = self._dataset.estimated_num_seqs
+        self._data_iter: Optional[Iterator[Tuple[int, TensorDict]]] = None
+        self._tensor_dict_template: Optional[TensorDict]
+
+        if self._map_outputs is not None:
+            self._tensor_dict_template = TensorDict(self._map_outputs)
+        else:
+            self._tensor_dict_template = TensorDict(
+                {name: self._make_tensor_template(name) for name in self._dataset.get_data_keys()}
+            )
```

Comment:
This code basically should always be done first like this:
```suggestion
            self._in_tensor_dict_template = TensorDict(
                {name: self._make_tensor_template_from_input(name) for name in self._dataset.get_data_keys()}
            )
```
And then if `map_outputs` is not defined, you do:
```python
self._out_tensor_dict_template = self._in_tensor_dict_template
```